### PR TITLE
WoRMS errors are caught and handled

### LINKED ISSUE #441: Handle WoRMS errors more gracefully
WoRMS is in maintenance mode at the moment. This produces a `SoapFault` error [here](https://github.com/biigle/core/blob/557bfb2fdf1d083cf2fe44e87796d25e8428b598/app/Services/LabelSourceAdapters/WormsAdapter.php#L91), which is not caught. Handle these kinds of errors more gracefully and return a "WoRMS is currently unavailable" response (maybe even code 503?).

### PR DESCRIPTION
**Pull request by lehecht**
Resolves #441.
SoapFault exception is caught in WormsAdapter and ServiceUnavailableHttpException is thrown instead. Tests were added to WormsAdapterTest.

### CODE REVIEW

File: app/Services/LabelSourceAdapters/WormsAdapter.php [hunk 1]
```diff
@@ -137,27 +146,34 @@ public function create($id, Request $request)
...
+                return $this->createRecursiveLabels($attributes);
+            }
+
+        } catch(\SoapFault $sf) {
+            throw new ServiceUnavailableHttpException(Response::HTTP_SERVICE_UNAVAILABLE, 'The label source WoRMS is currently unavailable.');
         }
 
-        return [$this->createSingleLabel($attributes)];
+            return [$this->createSingleLabel($attributes)];
+
```
Comment (mzur) [inline app/Services/LabelSourceAdapters/WormsAdapter.php:176]: Please fix the indentation and remove the empty line.

File: tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php [hunk 2]
```diff
@@ -119,6 +121,28 @@ public function testFindEmpty()
...
+
+        $request = new Request;
+        $request->merge(['query' => 'Kolga']);
+
+        try {
+            $adapter->find($request);
+        } catch(ServiceUnavailableHttpException $e) {
+            $this->assertEquals(Response::HTTP_SERVICE_UNAVAILABLE, $e->getStatusCode());
+            $this->assertEquals('test', $e->getMessage());
+        }
```
Comment (mzur) [inline tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php:143]: The test would pass if no error was thrown, too. In cases like these, I do:

```suggestion
        try {
            $adapter->find($request);
            $this->assertFalse(true);
        } catch(ServiceUnavailableHttpException $e) {
            $this->assertEquals('The WoRMS server is currently unavailable.', $e->getMessage());
        }
```

File: tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php [hunk 3]
```diff
@@ -170,6 +194,30 @@ public function testCreateNormal()
...
+        $request = new Request;
+        $request->merge(['query' => 'Kolga',
+                        'source_id' => 124731000,]);
+
+        try {
+            $adapter->create($tree->id, $request);
+        } catch(ServiceUnavailableHttpException $e) {
+            $this->assertEquals(Response::HTTP_SERVICE_UNAVAILABLE, $e->getStatusCode());
+            $this->assertEquals('test', $e->getMessage());
+        }
```
Comment (mzur) [inline tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php:218]:
```suggestion
        try {
            $adapter->create($tree->id, $request);
            $this->assertFalse(true);
        } catch(ServiceUnavailableHttpException $e) {
            $this->assertEquals('The WoRMS server is currently unavailable.', $e->getMessage());
        }
```

File: app/Services/LabelSourceAdapters/WormsAdapter.php [hunk 4]
```diff
@@ -137,27 +146,34 @@ public function create($id, Request $request)
...
+                return $this->createRecursiveLabels($attributes);
+            }
+
+        } catch (\SoapFault $sf) {
+            throw new ServiceUnavailableHttpException(message: 'The WoRMS server is currently unavailable.');
         }
 
         return [$this->createSingleLabel($attributes)];
+
```
Comment (mzur) [inline app/Services/LabelSourceAdapters/WormsAdapter.php:176]:
```suggestion
```

File: tests/php/Http/Controllers/Api/LabelSourceControllerTest.php [hunk 5]
```diff
@@ -8,6 +8,7 @@
 use Illuminate\Http\Request;
 use Illuminate\Http\Response;
 use Mockery;
+use Symfony\Component\HttpKernel\Exception\ServiceUnavailableHttpException;
```
Comment (mzur) [inline tests/php/Http/Controllers/Api/LabelSourceControllerTest.php:11]:
```suggestion
```

File: tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php [hunk 6]
```diff
@@ -119,6 +121,28 @@ public function testFindEmpty()
         $this->assertEquals([], $results);
     }
 
+    public function testFindThrowException()
+    {
+
+        $mock = Mockery::mock(SoapClient::class);
+        $mock->shouldReceive('getAphiaRecords')
+            ->once()
+            ->andThrow(new \SoapFault(Response::HTTP_SERVICE_UNAVAILABLE, 'test'));
```
Comment (mzur) [inline tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php:130]:
```suggestion
            ->andThrow(new \SoapFault);
```

File: tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php [hunk 7]
```diff
@@ -170,6 +194,30 @@ public function testCreateNormal()
...
     }
 
+    public function testCreateThrowException()
+    {
+        $tree = LabelTreeTest::create();
+
+        $mock = Mockery::mock(SoapClient::class);
+        $mock->shouldReceive('getAphiaNameByID')
+            ->once()
+            ->andThrow(new \SoapFault(Response::HTTP_SERVICE_UNAVAILABLE, 'test'));
```
Comment (mzur) [inline tests/php/Services/LabelSourceAdapters/WormsAdapterTest.php:204]:
```suggestion
            ->andThrow(new \SoapFault);
```

**Review by mzur — APPROVED** [review]
Thanks! There are only a few minor things left which I can easily apply myself in this PR.

One thing I noticed during PR review is that I open all the conversations again that you resolved (so I know what I said before). So maybe you can just leave them open and I close them when I review the new changes?

**Comment by lehecht** [comment]
> Thanks! There are only a few minor things left which I can easily apply myself in this PR.
> 
> One thing I noticed during PR review is that I open all the conversations again that you resolved (so I know what I said before). So maybe you can just leave them open and I close them when I review the new changes?

Sorry! I thought resolving those conversations is just a client thing. I will not change them for the next issues.

**Pull request merged**